client-core/src/wallet/rpc_sled_wallet_client.rs

```diff
@@ -0,0 +1,213 @@
+#![cfg(all(feature = "sled", feature = "rpc"))]
+
+use std::path::Path;
+
+use failure::ResultExt;
+use zeroize::Zeroize;
+
+use chain_core::init::address::RedeemAddress;
+use chain_core::init::coin::{sum_coins, Coin};
+use chain_core::tx::data::address::ExtendedAddr;
+use client_common::balance::TransactionChange;
+use client_common::storage::SledStorage;
+use client_common::{ErrorKind, Result};
+#[cfg(not(test))]
+use client_index::index::RpcSledIndex;
+use client_index::Index;
+
+use crate::service::*;
+#[cfg(test)]
+use crate::tests::MockIndex;
+use crate::{PublicKey, WalletClient};
+
+/// Wallet client backed by `sled` embedded database and `RpcSledIndex`
+pub struct RpcSledWalletClient {
+    key_service: KeyService<SledStorage>,
+    wallet_service: WalletService<SledStorage>,
+    #[cfg(not(test))]
+    index: RpcSledIndex,
+    #[cfg(test)]
+    index: MockIndex,
+}
+
+impl RpcSledWalletClient {
+    /// Creates a new instance of `RpcSledWalletClient`
+    pub fn new<P: AsRef<Path>>(path: P, url: &str) -> Result<Self> {
+        #[cfg(not(test))]
+        let storage = SledStorage::new(path.as_ref().to_path_buf())?;
+        #[cfg(test)]
+        let storage = SledStorage::temp(path.as_ref().to_path_buf())?;
+
+        #[cfg(not(test))]
+        let index = RpcSledIndex::new(path, url)?;
+        #[cfg(test)]
+        let index = MockIndex::new(path, url);
+
+        Ok(RpcSledWalletClient {
+            key_service: KeyService::new(storage.clone()),
+            wallet_service: WalletService::new(storage),
+            index,
+        })
+    }
+}
+
+impl WalletClient for RpcSledWalletClient {
+    fn new_wallet(&self, name: &str, passphrase: &str) -> Result<String> {
+        self.wallet_service.create(name, passphrase)
+    }
+
+    fn public_keys(&self, name: &str, passphrase: &str) -> Result<Vec<PublicKey>> {
+        let wallet_id = self.wallet_service.get(name, passphrase)?;
+
+        match wallet_id {
+            None => Err(ErrorKind::WalletNotFound.into()),
```

Comment:
Is it possible for us to distinguish if it is incorrect passphase or missing wallet?

Reply:
Yes.
`get_secure()` returns `DecryptionError` when passphrase is incorrect and `None` in cases where no value corresponding to a particular key is found ([code](https://github.com/crypto-com/chain/blob/master/client-common/src/storage.rs#L87)). Here, we're converting `None` into error of type `WalletNotFound`.